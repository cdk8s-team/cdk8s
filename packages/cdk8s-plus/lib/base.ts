import { Construct } from 'constructs';
import * as cdk8s from 'cdk8s';
import { ObjectMeta } from './object-meta';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   */
  readonly metadata?: ObjectMeta
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;
}

/**
 * Base class for all Kubernetes objects in stdk8s. Represents a single
 * resource.
 */
export abstract class Resource extends Construct implements IResource {

  /**
   * The metadata associated with this resource.
   */
  protected readonly metadata?: ObjectMeta;

  constructor(scope: Construct, id: string, props: ResourceProps) {
    super(scope, id);

    this.metadata = props.metadata;
  }

  /**
   * The underlying cdk8s API object.
   */
  public abstract readonly apiObject: cdk8s.ApiObject;

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }
}