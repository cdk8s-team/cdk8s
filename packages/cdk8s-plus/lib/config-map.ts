import { Construct } from 'constructs';
import * as k8s from '../imports/k8s';
import { ResourceProps, Resource, IResource } from './base';
import * as cdk8s from 'cdk8s';
import * as fs from 'fs';
import * as path from 'path';
import * as minimatch from 'minimatch';
import { onSynth, undefinedIfEmpty } from './utils';

/**
 * Initialization props for config maps.
 */
export interface ConfigMapProps extends ResourceProps {
  /**
   * BinaryData contains the binary data.
   *
   * Each key must consist of alphanumeric characters, '-', '_' or '.'.
   * BinaryData can contain byte sequences that are not in the UTF-8 range. The
   * keys stored in BinaryData must not overlap with the ones in the Data field,
   * this is enforced during validation process. Using this field will require
   * 1.10+ apiserver and kubelet.
   *
   * You can also add binary data using `configMap.addBinaryData()`.
   */
  readonly binaryData?: { [key: string]: string };

  /**
   * Data contains the configuration data.
   *
   * Each key must consist of alphanumeric characters, '-', '_' or '.'. Values
   * with non-UTF-8 byte sequences must use the BinaryData field. The keys
   * stored in Data must not overlap with the keys in the BinaryData field, this
   * is enforced during validation process.
   *
   * You can also add data using `configMap.addData()`.
   */
  readonly data?: { [key: string]: string };
}

/**
 * Represents a config map.
 */
export interface IConfigMap extends IResource {

}

/**
 * ConfigMap holds configuration data for pods to consume.
 */
export class ConfigMap extends Resource implements IConfigMap {
  /**
   * Represents a ConfigMap created elsewhere.
   * @param name The name of the config map to import
   */
  public static fromConfigMapName(name: string): IConfigMap {
    return { name };
  }

  public readonly apiObject: cdk8s.ApiObject;

  private readonly binaryData: { [key: string]: string } = { };
  private readonly data: { [key: string]: string } = { };

  public constructor(scope: Construct, id: string, props: ConfigMapProps = { }) {
    super(scope, id, props);

    this.apiObject = new k8s.ConfigMap(this, 'ConfigMap', {
      metadata: {
        name: this.metadata?.name,
        ...this.metadata?._toKube(),
      },
      data: onSynth(() => this.synthesizeData()),
      binaryData: onSynth(() => this.synthesizeBinaryData()),
    });

    for (const [k,v] of Object.entries(props.data ?? { })) {
      this.addData(k, v);
    }

    for (const [k,v] of Object.entries(props.binaryData ?? { })) {
      this.addBinaryData(k, v);
    }
  }

  /**
   * Adds a data entry to the config map.
   * @param key The key
   * @param value The value
   *
   * @throws if there is either a `data` or `binaryData` entry with the same key
   */
  public addData(key: string, value: string) {
    this.verifyKeyAvailable(key);

    this.data[key] = value;
  }

  /**
   * Adds a binary data entry to the config map. BinaryData can contain byte
   * sequences that are not in the UTF-8 range.
   * @param key The key
   * @param value The value
   *
   * @throws if there is either a `data` or `binaryData` entry with the same key
   */
  public addBinaryData(key: string, value: string) {
    this.verifyKeyAvailable(key);

    this.binaryData[key] = value;
  }

  /**
   * Adds a file to the ConfigMap.
   * @param localFile The path to the local file
   * @param key The ConfigMap key (default to the file name).
   */
  public addFile(localFile: string, key?: string) {
    key = key ?? path.basename(localFile);
    const value = fs.readFileSync(localFile, 'utf-8');

    this.addData(key, value);
  }

  /**
   * Adds a directory to the ConfigMap.
   * @param localDir A path to a local directory
   * @param options Options
   */
  public addDirectory(localDir: string, options: AddDirectoryOptions = { }) {
    if (options.recursive) {
      throw new Error("'recursive' is not supported (see https://github.com/kubernetes/kubernetes/pull/63362)");
    }

    const exclude = options.exclude ?? [];
    const shouldInclude = (file: string) => {
      for (const pattern of exclude) {
        if (minimatch(file, pattern)) {
          return false;
        }
      }
      return true;
    };

    const keyPrefix = options.keyPrefix ?? '';
    for (const file of fs.readdirSync(localDir)) {
      if (!shouldInclude(file)) {
        continue;
      }

      const filePath = path.join(localDir, file);
      const relativeFilePath = keyPrefix + file;
      if (options.recursive && fs.statSync(filePath).isDirectory()) {
        this.addDirectory(filePath, {
          keyPrefix: relativeFilePath + '/',
        });
      } else {
        this.addFile(filePath, relativeFilePath);
      }
    }
  }

  private verifyKeyAvailable(key: string) {
    if (key in this.data || key in this.binaryData) {
      throw new Error(`unable to add a ConfigMap entry with key "${key}". It is already used`);
    }
  }

  private synthesizeData() {
    return undefinedIfEmpty(this.data);
  }

  private synthesizeBinaryData() {
    return undefinedIfEmpty(this.binaryData);
  }
}


/**
 * Options for `configmap.addDirectory()`
 */
export interface AddDirectoryOptions {
  /**
   * A prefix to add to all keys in the config map.
   * @default ""
   */
  readonly keyPrefix?: string;

  /**
   * Glob patterns to exclude when adding files.
   * @default - include all files
   */
  readonly exclude?: string[];

  /**
   * Whether to descend to subdirectories (not supported yet).
   * @default false
   */
  readonly recursive?: boolean;
}
