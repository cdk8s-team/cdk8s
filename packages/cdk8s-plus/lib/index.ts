export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './duration';
export * from './job';
export * from './object-meta';
export * from './pod-template';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume-mount';
export * from './volume';