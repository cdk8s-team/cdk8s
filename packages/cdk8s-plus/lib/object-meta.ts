import * as k8s from '../imports/k8s';

export interface ObjectMetaProps {

  readonly annotations?: { [key: string]: string };

  readonly clusterName?: string;

  readonly labels?: { [key: string]: string };

  readonly name?: string;

  readonly namespace?: string;

}

export class ObjectMeta {

  public readonly clusterName?: string;
  public readonly name?: string;
  public readonly namespace?: string;
  public readonly labels: { [key: string]: string };
  public readonly annotations: { [key: string]: string };

  constructor(props: ObjectMetaProps = {}) {
    this.annotations = props.annotations ?? {};
    this.labels = props.labels ?? {};
    this.clusterName = props.clusterName;
    this.name = props.name;
    this.namespace = props.namespace;
  }

  public addLabel(key: string, value: string) {
    this.labels[key] = value;
  }

  public addAnnotation(key: string, value: string) {
    this.annotations[key] = value;
  }

  /**
   * @internal
   */
  public _toKube(): k8s.ObjectMeta {
    return {
      annotations: this.annotations,
      clusterName: this.clusterName,
      labels: this.labels,
      name: this.name,
      namespace: this.namespace,
    };
  }

}