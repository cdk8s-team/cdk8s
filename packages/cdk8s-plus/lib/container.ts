import { IConfigMap } from './config-map';
import { ISecret } from './secret';
import * as k8s from '../imports/k8s';
import { VolumeMount } from './volume-mount';

export interface ContainerProps {
  readonly image: string;

  /**
   * @default "main"
   */
  readonly name?: string;

  /**
   * // TODO: make this an array of structs (see k8s#ContainerPort)
   * @default - on port is exposed
   */
  readonly port?: number

  /**
   * The command to execute
   */
  readonly command?: string[];

  /**
   * Container's working directory.
   *
   * @default - If not specified, the container runtime's default will be used,
   * which might be configured in the container image. Cannot be updated.
   */
  readonly workingDir?: string;

  readonly env?: { [name: string]: EnvValue }
}

export interface EnvValueFromConfigMapOptions {
  /**
   * Specify whether the ConfigMap or its key must be defined
   * @default false
   */
  readonly optional?: boolean;
}

export interface EnvValueFromSecretOptions {

  /**
   * Specify whether the Secret or its key must be defined
   *
   * @default false
   */
  readonly optional?: boolean;

}

export interface EnvValueFromProcessOptions {

  readonly key: string;

  readonly required?: boolean;
}


export class EnvValue {
  public static of(value: string): EnvValue  {
    return { value };
  }

  public static fromConfigMap(configMap: IConfigMap, key: string, options: EnvValueFromConfigMapOptions = { }): EnvValue {
    return {
      valueFrom: {
        configMapKeyRef: {
          name: configMap.name,
          key,
          optional: options.optional,
        },
      } as k8s.EnvVarSource,
    };
  }

  public static fromSecret(secret: ISecret, key: string, options: EnvValueFromSecretOptions = {}): EnvValue {
    return {
      valueFrom: {
        secretKeyRef: {
          name: secret.name,
          key,
          optional: options.optional,
        },
      } as k8s.EnvVarSource,
    };

  }

  public static fromValue(value: string): EnvValue {
    return {
      value: value,
    };
  }

  // TODO: not sure about this. lets see how it feels.
  public static fromProcess(options: EnvValueFromProcessOptions): EnvValue {

    const value = process.env[options.key];

    if (options.required && !value) {
      throw new Error(`Missing ${options.key} env variable`);
    }

    return EnvValue.fromValue(value!);
  }

  private constructor(public readonly value?: any, public readonly valueFrom?: any) {

  }
}

export class Container {
  public readonly volumeMounts: VolumeMount[] = [];
  public readonly name: string;
  public readonly image: string;
  public readonly port?: number;
  public readonly command?: string[];
  public readonly workingDir?: string;
  public readonly env: { [name: string]: EnvValue };

  constructor(props: ContainerProps) {
    this.name = props.name ?? 'main';
    this.image = props.image;
    this.port = props.port;
    this.command = props.command;
    this.env = props.env ?? { };
    this.workingDir = props.workingDir;
  }

  public addEnv(name: string, value: EnvValue) {
    this.env[name] = value;
  }

  public mount(options: VolumeMount) {
    this.volumeMounts.push(options);
  }
}