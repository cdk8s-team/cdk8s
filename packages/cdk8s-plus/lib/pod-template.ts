import * as k8s from '../imports/k8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import * as cdk8s from 'cdk8s';
import { onSynth } from './utils';
import { ObjectMeta } from './object-meta';
import { PodSpec } from './pod';

export interface PodTemplateProps extends ResourceProps {

  readonly spec: PodTemplateSpec;

}

export class PodTemplate extends Resource {

  public readonly apiObject: cdk8s.ApiObject;

  private readonly spec: PodTemplateSpec;

  constructor(scope: Construct, name: string, props: PodTemplateProps) {
    super(scope, name, props);

    this.spec = props.spec;

    this.apiObject = new k8s.PodTemplate(this, 'PodTemplate', {
      metadata: {
        name: this.metadata?.name,
        ...this.metadata?._toKube(),
      },
      template: this.spec._toKube(),
    })

  }


}

export interface PodTemplateSpecProps {

  readonly metadata?: ObjectMeta;

  readonly podSpec?: PodSpec;

}

export class PodTemplateSpec {

  public readonly metadata: ObjectMeta;
  public readonly podSpec: PodSpec;

  constructor(props: PodTemplateSpecProps = {}) {
    this.metadata = props.metadata ?? new ObjectMeta();
    this.podSpec = props.podSpec ?? new PodSpec();
  }

  /**
   * @internal
   */
  public _toKube(): k8s.PodTemplateSpec {
    return onSynth(() => ({
      metadata: this.metadata._toKube(),
      spec: this.podSpec._toKube(),
    }));
  }

}