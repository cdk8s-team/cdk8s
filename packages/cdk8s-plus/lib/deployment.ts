import * as k8s from '../imports/k8s';
import { Construct, Node } from 'constructs';
import { Service, ServiceSpec } from './service';
import { Resource, ResourceProps } from './base';
import * as cdk8s from 'cdk8s';
import { onSynth } from './utils';
import { PodTemplateSpec } from './pod-template';

export interface DeploymentProps extends ResourceProps {

  readonly spec?: DeploymentSpec;

}

export interface ExposeOptions {

  readonly port: number;
}

export class Deployment extends Resource {

  public readonly apiObject: cdk8s.ApiObject;

  public readonly spec: DeploymentSpec;

  constructor(scope: Construct, id: string, props: DeploymentProps = {}) {
    super(scope, id, props);

    this.spec = props.spec ?? new DeploymentSpec();

    this.apiObject = new k8s.Deployment(this, 'Deployment', {
      metadata: {
        name: this.metadata?.name,
        ...this.metadata?._toKube(),
      },
      spec: this.spec._toKube(Node.of(this)),
    })
  }

  public expose(options: ExposeOptions): Service {

    // just a PoC, we assume the first container is the main one.
    // TODO: figure out what the correct thing to do here.
    const containerPort = this.spec.template.podSpec.containers[0].port;

    // create a label and attach it to the deployment pods
    const labelKey = 'cdk8s.deployment'
    const labelValue = Node.of(this).uniqueId

    // TODO: what are the implications of using 'this' as the scope? should we accept as argument?
    const service = new Service(this, 'Service', {
      spec: new ServiceSpec({
        port: {
          port: options.port,
          targetPort: containerPort,
        },
      }),
    })

    service.spec.selectByLabel(labelKey, labelValue);

    return service;
  }

}

export interface DeploymentSpecProps {

  readonly replicas?: number;

  readonly template?: PodTemplateSpec;

}

export class DeploymentSpec {

  private readonly replicas?: number;

  public readonly template: PodTemplateSpec;
  public readonly labels: Record<string, string>;

  constructor(props: DeploymentSpecProps = {}) {
    this.replicas = props.replicas;
    this.template = props.template ?? new PodTemplateSpec();
    this.labels = {};
  }

  public selectByLabel(key: string, value: string) {
    this.labels[key] = value;
  }

  /**
   * @internal
   */
  public _toKube(node: Node): k8s.DeploymentSpec {

    // automatically select pods in this deployment

    const selector = 'cdk8s.deployment';
    const matcher = node.uniqueId;

    this.template.metadata.addLabel(selector, matcher);

    this.selectByLabel(selector, matcher);

    return onSynth(() => ({
      replicas: this.replicas,
      template: this.template._toKube(),
      selector: {
        matchLabels: this.labels,
      },
    }));
  }
}
