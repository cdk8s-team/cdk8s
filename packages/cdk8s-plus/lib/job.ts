import { Resource, ResourceProps } from './base';
import { ApiObject } from 'cdk8s';
import { Construct } from 'constructs';

import * as k8s from '../imports/k8s';
import { PodTemplateSpec } from './pod-template';
import { RestartPolicy } from './pod';
import { Duration } from './duration';


export interface JobProps extends ResourceProps {
  readonly spec: JobSpec;
}

export class Job extends Resource {
  public readonly apiObject: ApiObject;
  public readonly spec: JobSpec;

  constructor(scope: Construct, id: string, props: JobProps) {
    super(scope, id, props);

    this.spec = props.spec;

    this.apiObject = new k8s.Job(this, 'Default', {
      metadata: {
        name: this.metadata?.name,
        ...this.metadata?._toKube(),
      },
      spec: this.spec._toKube(),
    });
  }
}

export interface JobSpecProps {
  readonly template: PodTemplateSpec;
  /**
   * Limits the lifetime of a Job that has finished execution (either Complete
   * or Failed). If this field is set, after the Job finishes, it is eligible to
   * be automatically deleted. When the Job is being deleted, its lifecycle
   * guarantees (e.g. finalizers) will be honored. If this field is set to zero,
   * the Job becomes eligible to be deleted immediately after it finishes. This
   * field is alpha-level and is only honored by servers that enable the
   * `TTLAfterFinished` feature.
   *
   * @default - If this field is unset, the Job won't be automatically deleted.
   */
  readonly ttlAfterFinished?: Duration;
}

export class JobSpec {

  public readonly template: PodTemplateSpec;
  public readonly ttlAfterFinished?: Duration;

  constructor(props: JobSpecProps) {
    this.template = props.template;
    this.ttlAfterFinished = props.ttlAfterFinished;

    if (!this.template.podSpec.restartPolicy) {
      this.template.podSpec.restartPolicy = RestartPolicy.NEVER;
    }
  }

  /**
   * @internal
   */
  public _toKube(): k8s.JobSpec {
    return {
      template: this.template._toKube(),
      ttlSecondsAfterFinished: this.ttlAfterFinished ? this.ttlAfterFinished.toSeconds() : undefined,
    };
  }
}