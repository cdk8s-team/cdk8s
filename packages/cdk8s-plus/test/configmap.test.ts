import { ConfigMap, ObjectMeta } from '../lib';
import { Testing } from 'cdk8s';

test('minimal', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  new ConfigMap(chart, 'my-config-map');

  // THEN
  expect(Testing.synth(chart)).toStrictEqual([{
    'apiVersion': 'v1',
    'kind': 'ConfigMap',
    'metadata': {
      'name': 'test-my-config-map-configmap-d0fa5644',
    },
  }]);
});

test('with data', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  new ConfigMap(chart, 'my-config-map', {
    data: {
      key1: 'foo',
      key2: 'bar',
    },
  });

  // THEN
  expect(Testing.synth(chart)).toStrictEqual([{
    'apiVersion': 'v1',
    'kind': 'ConfigMap',
    'data': {
      'key1': 'foo',
      'key2': 'bar',
    },
    'metadata': {
      'name': 'test-my-config-map-configmap-d0fa5644',
    },
  }]);
});

test('with binaryData', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  new ConfigMap(chart, 'my-config-map', {
    binaryData: {
      key1: 'foo',
      key2: 'bar',
    },
  });

  // THEN
  expect(Testing.synth(chart)).toStrictEqual([{
    'apiVersion': 'v1',
    'kind': 'ConfigMap',
    'binaryData': {
      'key1': 'foo',
      'key2': 'bar',
    },
    'metadata': {
      'name': 'test-my-config-map-configmap-d0fa5644',
    },
  }]);
});

test('with binaryData and data', () => {
  // GIVEN
  const chart = Testing.chart();

  // WHEN
  new ConfigMap(chart, 'my-config-map', {
    data: {
      hello: 'world',
    },
    binaryData: {
      key1: 'foo',
      key2: 'bar',
    },
  });

  // THEN
  expect(Testing.synth(chart)).toStrictEqual([{
    'apiVersion': 'v1',
    'kind': 'ConfigMap',
    'data': {
      'hello': 'world',
    },
    'binaryData': {
      'key1': 'foo',
      'key2': 'bar',
    },
    'metadata': {
      'name': 'test-my-config-map-configmap-d0fa5644',
    },
  }]);
});

test('"binaryData" and "data" cannot share keys', () => {
  // GIVEN
  const chart = Testing.chart();

  // THEN
  expect(() => new ConfigMap(chart, 'my-config-map', {
    data: {
      key1: 'world',
    },
    binaryData: {
      key1: 'foo',
      key2: 'bar',
    },
  })).toThrow(/unable to add a ConfigMap entry with key \"key1\". It is already used/);
});

test('addData()/addBinaryDataq() can be used to add data', () => {
  // GIVEN
  const chart = Testing.chart();
  const cm = new ConfigMap(chart, 'my-config-map', {
    data: {
      hello: 'world',
    },
    binaryData: {
      key1: 'foo',
      key2: 'bar',
    },
  });

  // WHEN
  cm.addData('world', 'oh yeah!');
  cm.addBinaryData('key3', 'baz');

  // THEN
  expect(Testing.synth(chart)).toStrictEqual([{
    'apiVersion': 'v1',
    'binaryData': {
      'key1': 'foo',
      'key2': 'bar',
      'key3': 'baz',
    },
    'data': {
      'hello': 'world',
      'world': 'oh yeah!',
    },
    'kind': 'ConfigMap',
    'metadata': {'name': 'test-my-config-map-configmap-d0fa5644'},
  }]);
});

test('addData() and addBinaryData() throw if key already used', () => {
  // GIVEN
  const chart = Testing.chart();
  const cm = new ConfigMap(chart, 'my-config-map', {
    data: {
      key: 'value',
    },
  });

  // THEN
  expect(() => cm.addData('key', 'value2')).toThrow(/unable to add a ConfigMap entry with key \"key\". It is already used/);
  expect(() => cm.addBinaryData('key', 'value2')).toThrow(/unable to add a ConfigMap entry with key \"key\". It is already used/);
});

test('addFile() adds local files to the config map', () => {
  // GIVEN
  const chart = Testing.chart();
  const cm = new ConfigMap(chart, 'my-config-map');

  // WHEN
  cm.addFile(`${__dirname}/fixtures/file1.txt`); // defaults to file name
  cm.addFile(`${__dirname}/fixtures/file2.html`, 'hey-there');

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

describe('addDirectory() adds all files in a directory to the config map', () => {

  test('keys are based on file names', () => {
    // GIVEN
    const chart = Testing.chart();
    const cm = new ConfigMap(chart, 'my-config-map');

    // WHEN
    cm.addDirectory(`${__dirname}/fixtures`);

    // THEN
    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('"prefix" can be used to prefix keys"', () => {
    // GIVEN
    const chart = Testing.chart();
    const cm = new ConfigMap(chart, 'my-config-map');

    // WHEN
    cm.addDirectory(`${__dirname}/fixtures`, { keyPrefix: 'prefix.' });

    // THEN
    expect(Testing.synth(chart)).toMatchSnapshot();
  })

  test('"exclude" exclusion via globs', () => {
    // GIVEN
    const chart = Testing.chart();
    const cm = new ConfigMap(chart, 'my-config-map');

    // WHEN
    cm.addDirectory(`${__dirname}/fixtures`, { exclude: [ '*.html' ] });

    // THEN
    expect(Testing.synth(chart)).toMatchSnapshot();
  });

  test('"recursive" is not supported yet', () => {
    // GIVEN
    const chart = Testing.chart();
    const cm = new ConfigMap(chart, 'my-config-map');

    // WHEN
    expect(() => cm.addDirectory(`${__dirname}/fixtures`, { recursive: true })).toThrow(/'recursive' is not supported/);
  });
});

test('metadata is synthesized', () => {

  const chart = Testing.chart();
  new ConfigMap(chart, 'my-config-map', {
    metadata: new ObjectMeta({
      name: 'my-name',
    }),
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();

});

