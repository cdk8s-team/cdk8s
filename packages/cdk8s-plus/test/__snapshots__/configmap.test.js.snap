// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addDirectory() adds all files in a directory to the config map "exclude" exclusion via globs 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "file1.txt": "Hello, world!",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "test-my-config-map-configmap-d0fa5644",
    },
  },
]
`;

exports[`addDirectory() adds all files in a directory to the config map "prefix" can be used to prefix keys" 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "prefix.file1.txt": "Hello, world!",
      "prefix.file2.html": "<html>Hey</html>",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "test-my-config-map-configmap-d0fa5644",
    },
  },
]
`;

exports[`addDirectory() adds all files in a directory to the config map keys are based on file names 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "file1.txt": "Hello, world!",
      "file2.html": "<html>Hey</html>",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "test-my-config-map-configmap-d0fa5644",
    },
  },
]
`;

exports[`addFile() adds local files to the config map 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "file1.txt": "Hello, world!",
      "hey-there": "<html>Hey</html>",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "test-my-config-map-configmap-d0fa5644",
    },
  },
]
`;

exports[`metadata is synthesized 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ConfigMap",
    "metadata": Object {
      "annotations": Object {},
      "labels": Object {},
      "name": "my-name",
    },
  },
]
`;
